stdlib/public/runtime/DynamicCast.cpp

```diff
@@ -206,13 +206,86 @@ struct _ObjectiveCBridgeableWitnessTable : WitnessTable {
 extern "C" const ProtocolDescriptor
 PROTOCOL_DESCR_SYM(s21_ObjectiveCBridgeable);
 
+#if SWIFT_OBJC_INTEROP
+#define BRIDGING_CONFORMANCE_SYM \
+  MANGLE_SYM(s19_BridgeableMetatypeVs21_ObjectiveCBridgeablesWP)
+
+extern "C" const _ObjectiveCBridgeableWitnessTable BRIDGING_CONFORMANCE_SYM;
+#endif
+
+/// Nominal type descriptor for Swift.String.
+extern "C" const StructDescriptor NOMINAL_TYPE_DESCR_SYM(SS);
+
+struct ObjCBridgeWitnessCacheEntry {
+  const Metadata *metadata;
+  const _ObjectiveCBridgeableWitnessTable *witness;
+};
+
 static const _ObjectiveCBridgeableWitnessTable *
-findBridgeWitness(const Metadata *T) {
+swift_conformsToObjectiveCBridgeableNoCache(const Metadata *T) {
   auto w = swift_conformsToProtocolCommon(
-      T, &PROTOCOL_DESCR_SYM(s21_ObjectiveCBridgeable));
+         T, &PROTOCOL_DESCR_SYM(s21_ObjectiveCBridgeable));
   return reinterpret_cast<const _ObjectiveCBridgeableWitnessTable *>(w);
 }
 
+static const _ObjectiveCBridgeableWitnessTable *
+swift_conformsToObjectiveCBridgeable(const Metadata *T) {
+  static std::atomic<ObjCBridgeWitnessCacheEntry> _objcBridgeWitnessCache = {};
+  auto cached = _objcBridgeWitnessCache.load(SWIFT_MEMORY_ORDER_CONSUME);
+  if (cached.metadata == T) {
+    return cached.witness;
+  }
+  cached.witness = swift_conformsToObjectiveCBridgeableNoCache(T);
+  cached.metadata = T;
+  _objcBridgeWitnessCache.store(cached, std::memory_order_release);
+  return cached.witness;
+}
+
+static const _ObjectiveCBridgeableWitnessTable *
+findBridgeWitness(const Metadata *T) {
```

Comment:
This entire function is duplicated between DynamicCast.cpp and Casting.cpp, along with its memoization variables. There's also a separate memoization of String stuff in DynamicCast.cpp. I'm unifying the behavior of these (yes they had behavior differences), but saving actually unifying the implementations for later.